lib/recovery.c

```diff
@@ -0,0 +1,107 @@
+/*
+ * Microsemi Switchtec(tm) PCIe Management Library
+ * Copyright (c) 2017, Microsemi Corporation
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+/**
+ * @file
+ * @brief Switchtec core library functions for secure boot operations
+ */
+
+#include "switchtec_priv.h"
+#include "switchtec/switchtec.h"
+#include "switchtec/errors.h"
+#include "switchtec/endian.h"
+#include "switchtec/mrpc.h"
+#include "switchtec/errors.h"
+#include <unistd.h>
+
+#include <errno.h>
+#include <stdio.h>
+#include <string.h>
+
+#define IS_UNSUPPORTED_MRPC(x) ((x & (~SWITCHTEC_ERRNO_MRPC_FLAG_BIT)) == \
+				ERR_MPRC_UNSUPPORTED)
+
+static enum switchtec_gen map_to_gen(uint32_t gen)
+{
+	enum switchtec_gen ret = SWITCHTEC_GEN_UNKNOWN;
+
+	switch (gen) {
+	case 0:
+		ret = SWITCHTEC_GEN4;
+		break;
+	default:
+		ret = SWITCHTEC_GEN_UNKNOWN;
+		break;
+	}
+
+	return ret;
+}
+
+/**
+ * @brief Ping a device with a 32-bit data.
+ * @param[in]  dev	Switchtec device handle
+ * @param[in]  ping_dw	Ping data
+ * @param[out] reply_dw	Ping reply data (2's complement of ping data)
+ * @param[out] phase	The current boot phase
+ * @param[out] gen	Device generation
+ * @param[out] rev	Device revision
+ * @return 0 on success, error code on failure
+ */
+int switchtec_get_device_info(struct switchtec_dev *dev, uint32_t ping_dw,
+			      uint32_t *reply_dw,
```

Comment:
It seems like the ping_dw and reply_dw is not necessary as parameters if this API is to get the hw device info.